orttraining/orttraining/python/training/orttrainer_options.py

```diff
@@ -0,0 +1,416 @@
+                                   'gradient_accumulation_steps' : 128
+                               },
+                               'device' : {
+                                   'id' : 'cuda:0',
+                                   'mem_limit' : 2*1024*1024*1024,
+                               },
+                               'lr_scheduler' : optim.lr_scheduler.LinearWarmupLRScheduler(),
+                               'mixed_precision' : {
+                                   'enabled': True,
+                                   'loss_scaler': amp.LossScaler(loss_scale=float(1 << 16))
+                               }
+            })
+            fp16_enabled = opts.mixed_precision.enabled
+     """
+
+    def __init__(self, options):
+        # Keep a copy of original input for debug
+        self._original_opts = dict(options)
+
+        # Used for logging purposes
+        self._main_class_name = self.__class__.__name__
+
+        # Validates user input
+        self._validated_opts = dict(self._original_opts)
+        validator = ORTTrainerOptionsValidator(
+            _ORTTRAINER_OPTIONS_SCHEMA)
+        self._validated_opts = validator.validated(self._validated_opts)
+        if self._validated_opts is None:
+            raise ValueError(f'Invalid options: {validator.errors}')
+
+        # Convert dict in object
+        for k, v in self._validated_opts.items():
+            setattr(self, k, self._wrap(v))
+
+        # Keep this in the last line
+        # After this point, this class becomes immutable
+        self._initialized = True
+
+    def __repr__(self):
+        return '{%s}' % str(', '.join("'%s': %s" % (k, repr(v))
+                                      for (k, v) in self.__dict__.items()
+                                      if k not in ['_original_opts', '_validated_opts', '_initialized']))
+
+    def __setattr__(self, k, v):
+        if hasattr(self, '_initialized'):
+            raise Exception(f"{self._main_class_name} is an immutable class")
+        return super().__setattr__(k, v)
+
+    def _wrap(self, v):
+        if isinstance(v, (tuple, list, set, frozenset)):
+            return type(v)([self._wrap(v) for v in v])
+        else:
+            return _ORTTrainerOptionsInternal(self._main_class_name, v) if isinstance(v, dict) else v
+
+
+class _ORTTrainerOptionsInternal(ORTTrainerOptions):
+    r"""Internal class used by ONNX Runtime training backend for input validation
+
+    NOTE: Users MUST NOT use this class in any way!
+    """
+
+    def __init__(self, main_class_name, options):
+        # Used for logging purposes
+        self._main_class_name = main_class_name
+
+        # Convert dict in object
+        for k, v in dict(options).items():
+            setattr(self, k, self._wrap(v))
+
+        # Keep this in the last line
+        # After this point, this class becomes immutable
+        self._initialized = True
+
+
+class ORTTrainerOptionsValidator(cerberus.Validator):
```

Comment:
Do we want to depend on a third party package. Can we just use python typing?

Reply:
Ideally depending on Python standard would be ideal. But Python does not provide schema validation and implementing it from scratch is tricky and demands a lot of code. Cerberus can parse a dict recursively, enforce required parameters, check values for both required and optional field, can return default values when optional fields are not set, return a normalized dictionary with all not-set-but-with-existing-default-values and more. Not doing such validation would make user experience quite cumbersome, as unexpected crashes would have due to not validated input. An external dependency is a price to pay for it :-S

Reply:
I will create a separate task for investigating the replacement of cerberus by typing/dataclasses